### PR TITLE
Auto report unable to decrypt errors via lab option

### LINKED ISSUE #4263: [EA] Implement opt-in lab auto rageshake UISIs
### Your use case

See vector-im/element-meta#37

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #4263

Adds a new lab option to enable auto reporting of fail to decrypt errors. When detected the device will send a rageshake and then send a toDevice message to the other participant to request him to also upload a RS.

image

The AutoRageShake and UISIs detection part has been put in the app side, and in order to do that I added to new services in the SDK:

 - SendToDeviceService: A service that allows to send to device messages (needed by AutoRageshaker in order to ask for the matching rageshake)
 - EventStreamService: Provides entry points needed by the UISI detector (could not be done by listening at the db events..)
 
@ganfra / @bmarty  If you could have a quick look on the SDK part and give some feedbaks

Sample report 
https://github.com/matrix-org/element-android-rageshakes/issues/27579

### CODE REVIEW

**Comment by bmarty** [comment]
> @ganfra / @bmarty If you could have a quick look on the SDK part and give some feedbaks

Will do, I was waiting for the PR to not be draft to just read the description :/

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultToDeviceService.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import org.matrix.android.sdk.api.session.events.model.EventType
+import org.matrix.android.sdk.internal.crypto.actions.MessageEncrypter
+import org.matrix.android.sdk.internal.crypto.model.MXUsersDevicesMap
+import org.matrix.android.sdk.internal.crypto.store.IMXCryptoStore
+import org.matrix.android.sdk.internal.crypto.tasks.DefaultSendToDeviceTask
+import org.matrix.android.sdk.internal.crypto.tasks.SendToDeviceTask
+import javax.inject.Inject
+
+internal class DefaultToDeviceService @Inject constructor(
+        private val sendToDeviceTask: DefaultSendToDeviceTask,
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultToDeviceService.kt:30]: The type must be `SendToDeviceTask`

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/SessionModule.kt [hunk 2]
```diff
@@ -387,4 +389,10 @@ internal abstract class SessionModule {
 
     @Binds
     abstract fun bindEventSenderProcessor(processor: EventSenderProcessorCoroutine): EventSenderProcessor
+
+    @Binds
+    abstract fun bindToDeviceService(deviceService: DefaultToDeviceService): ToDeviceService
+
+    @Binds
+    abstract fun bindEventStreamService(deviceService: DefaultEventStreamService): EventStreamService
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/SessionModule.kt:397]: Could be grouped with the other services. Also rename param to `service` (at 2 places)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/StreamEventsManager.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+@SessionScope
+internal class StreamEventsManager @Inject constructor() {
+
+    private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
+
+    private val listeners = mutableListOf<LiveEventListener>()
+
+    fun addLiveEventListener(listener: LiveEventListener) {
+        Timber.v("## VALR: addLiveEventListener")
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/StreamEventsManager.kt:38]: use logger tag instead of VALR :)

**Review by bmarty — COMMENTED** [review]
Some small first remarks.
Weird to expose a ToDeviceService, but why not...

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/SessionModule.kt (hunk 2 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/SessionModule.kt]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultToDeviceService.kt (hunk 1 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultToDeviceService.kt]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/StreamEventsManager.kt (hunk 3 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/StreamEventsManager.kt]: :) Updated, just a temporary log that I forgot to remove

File: vector/src/main/java/im/vector/app/AutoRageShaker.kt [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+                                        userId = target.senderUserId,
+                                        deviceId = target.senderDeviceId,
+                                        content = mapOf(
+                                                "event_id" to target.eventId,
+                                                "room_id" to target.roomId,
+                                                "session_id" to target.sessionId,
+                                                "device_id" to target.senderDeviceId,
+                                                "user_id" to target.senderUserId,
+                                                "sender_key" to target.senderKey,
+                                                "matching_issue" to reportUrl
```
Comment (duxovni) [inline vector/src/main/java/im/vector/app/AutoRageShaker.kt:188]: `matching_issue` is a confusing name for this field, I'd expect that to be a link to a github issue or something; how would you feel about something more like `recipient_rageshake`?

Comment (duxovni) [reply vector/src/main/java/im/vector/app/AutoRageShaker.kt:188]: Sorry, didn't realize the rageshake server actually does create a github issue and responds with that; I'd had that disabled in my tests, and assumed we were looking at the URL of the submitted rageshake.

File: vector/src/main/res/xml/vector_settings_labs.xml [hunk 5]
```diff
@@ -44,11 +44,6 @@
         android:key="SETTINGS_LABS_UNREAD_NOTIFICATIONS_AS_TAB"
         android:title="@string/labs_show_unread_notifications_as_tab" />
 
-    <im.vector.app.core.preference.VectorSwitchPreference
-        android:defaultValue="false"
-        android:key="SETTINGS_LABS_USE_RESTRICTED_JOIN_RULE"
-        android:summary="@string/labs_use_restricted_join_rule_desc"
-        android:title="@string/labs_use_restricted_join_rule" />
```
Comment (bmarty) [inline vector/src/main/res/xml/vector_settings_labs.xml:51]: This removal will need more cleanup in the code, and not really related to this PR.

**Review by bmarty — APPROVED** [review]
LGTM

**Review by ganfra — APPROVED** [review]
Nice work Valere! I don't see any issue like that. Performance wise, it should be ok I guess as long as there is not too many listeners on EventStreamService. 
Other way of doing that is exposing Flows of event.
Maybe we should at least make them suspend methods? (LiveEventListener)

File: vector/src/main/res/xml/vector_settings_labs.xml (hunk 5 above)
Comment (BillCarsonFr) [reply vector/src/main/res/xml/vector_settings_labs.xml]: Agreed will restore and creating separate PR

Comment (BillCarsonFr) [reply vector/src/main/res/xml/vector_settings_labs.xml]: PR up here https://github.com/vector-im/element-android/pull/4889

**Pull request merged**